File: src/examples/chapter-15/client/next/src/domain/models/error/domain-error.ts
// Domain error models

export class DomainError extends Error {
  constructor(message: string, public readonly code?: string) {
    super(message);
    this.name = 'DomainError';
  }
}

export class NetworkError extends DomainError {
  constructor(message: string, public readonly statusCode?: number) {
    super(message, 'NETWORK_ERROR');
    this.name = 'NetworkError';
  }
}

export class AuthError extends DomainError {
  constructor(message: string) {
    super(message, 'AUTH_ERROR');
    this.name = 'AuthError';
  }
}

File: src/examples/chapter-15/client/flutter/lib/domain/models/error/domain_error.dart
/// Base class for all domain errors in Recipe Quest
abstract class DomainError extends Error {
  DomainError(this.message);

  final String message;

  @override
  String toString() => '$runtimeType: $message';
}

/// Error related to network operations
class NetworkError extends DomainError {
  NetworkError(super.message);
}

/// Error related to recipe generation
class RecipeGenerationError extends DomainError {
  RecipeGenerationError(super.message);
}

/// Error related to image generation
class ImageGenerationError extends DomainError {
  ImageGenerationError(super.message);
}

/// Error related to dish evaluation
class EvaluationError extends DomainError {
  EvaluationError(super.message);
}

/// Error related to invalid input or validation
class ValidationError extends DomainError {
  ValidationError(super.message);
}

/// Error related to parsing or serialization
class ParseError extends DomainError {
  ParseError(super.message);
}
